### ISSUE TITLE
threadpool.setMaxPoolSize(1)

### ISSUE THREAD

**Issue by pb-cdunn** [nim-lang/Nim#5499]
`threadpool.setMaxPoolSize(1)` seems to be ignored. The number of threads seems to be the number of processors.

I see this by observation:

1. All threads start, and later all threads end.
2. I have a `threadvar` for testing, and it gets initialized N times.

The NPROC threads seem to be created in `PreMain()`. I think I would have to dive into the Nim compiler to debug this. Should I?

(Btw, is there a way to print the current ThreadId? That would help me.)

**Comment by pb-cdunn** [comment]
Some reasons why this is important:

1. For testing, I like to set the threadpool size to 1 (or maybe 0, relying on `spawnX`).
2. With `spawnX` it can make more sense for the size to be `nproc - 1`.
3. Basic control.

**Comment by pb-cdunn** [comment]
I see the issue. When `threadpool` is imported, it calls `setup()`, which creates NPROC threads (48 in my case). Changing the max actually only sets the shutdown flag on the excess workers. Those workers will still accept jobs, at least the first time.

Maybe they shutdown after their first completion. I'm not sure. But this whole thing is a red herring anyway. I was able to run with a threadpool-size of `1` by simply modifying `Nim/lib/pure/concurrency/threadpool.nim`. That did not fix my runtime problem. Things are simply much slower in a thread than they are with Python's **multiprocessing** module, calling the exact same C extension module with the exact same data. In fact, I think I've isolated the runtime to the C code itself. My best guess is "False Sharing", which causes a high number of cache misses. That's just a guess, but I can say for sure that the Nim+C code is a bit faster than the Python+C code when all is done in the main thread of a single process.

Anyway, you must have a reason for starting with NPROC threads, so I'll close this for now.
